library_management_system_downloader/copy_and_compress_database_and_extract_tsv.py

```diff
@@ -0,0 +1,146 @@
+
+# Initialize the database:
+Base = automap_base()
+sql_engine = create_engine(f'sqlite:///{sqlite_database_location}')
+Base.prepare(sql_engine, reflect=True)
+
+dois_table = Base.classes.dois_table
+library_holdings_data = Base.classes.library_holdings_data
+
+# From the sqlalchemy documentation
+# (http://docs.sqlalchemy.org/en/latest/orm/session_basics.html)
+# Create a configured "session":
+Session = sessionmaker(bind=sql_engine)
+sql_session = Session()
+
+# =============================================================================
+# Create the data directory if it doesn't already exist
+# =============================================================================
+
+Path(data_directory).mkdir(parents=True, exist_ok=True)
+
+# =============================================================================
+# Extract data from the databse, and save that as a TSV
+# =============================================================================
+
+# We'll first define a subquery to get just the latest full text information
+# for each DOI:
+dataset_max_timestamp_for_each_doi_query = sql_session.query(
+        func.max(library_holdings_data.primary_key)).group_by(
+                library_holdings_data.doi_foreign_key)
+
+# str(dataset_max_timestamp_for_each_doi_query)  # See the SQL being used.
+# list(dataset_max_timestamp_for_each_doi_query)  # See the list of max
+# primary keys.
+
+# We'll now get the DOI and full-text indicator based on the subquery above:
+dataset_join_query = sql_session.query(
+        dois_table.doi,
+        # library_holdings_data.timestamp,  # Uncomment to also get the
+        # timestamp of the DOI fulltext query
+        # library_holdings_data.xml_response,  # Uncomment to also get the
+        # full XML response.
+        library_holdings_data.full_text_indicator) \
+        .join(library_holdings_data) \
+        .group_by(dois_table.doi) \
+        .filter(library_holdings_data.primary_key.in_(
+                dataset_max_timestamp_for_each_doi_query))
+
+joined_dataset = pd.read_sql(
+            dataset_join_query.statement,
+            dataset_join_query.session.bind)
+
+# Sort the dataset by DOI and timestamp, so that it will always be in the
+# same order:
+joined_dataset.sort_values(
+        by=["doi"],
+        inplace=True)
+
+tsv_output_path = str(Path(data_directory, tsv_output_file_name))
+
+logging.info(
+        f'Creating a TSV from the database, at "{tsv_output_path}"...')
+
+joined_dataset.to_csv(
+        path_or_buf=tsv_output_path,
+        sep='\t',
+        index=False,  # Do not write row numbers
+        compression='xz')
+
+# =============================================================================
+# Save a compressed copy of the database
+# =============================================================================
+
+
+def xz_compress_a_file(
+        input_file_location,
+        output_file_location,
+        is_the_file_binary=True):
```

Comment:
I don't understand the need for `is_the_file_binary`... you always treat a file as binary when compressing it?

Reply:
Also this argument is entirely non-functional at the moment. [`lzma`](https://docs.python.org/3/library/lzma.html) defaults to binary mode... `r` is equivalent to `rb` and `w` is equivalent to `wb`.

Reply:
Since memory-efficient on-disk compression took us a while to figure out, I [opened an issue](https://bugs.python.org/issue32160) on Python's tracker. So maybe one day we can get our solution added to the official Python!

Reply:
There's a bug here -- fixed in 0a720d5. Does this make more sense? I meant to only use `is_the_file_binary` for the input file. I can't find where we originally discussed it, but as I remember, we talked about wrapping the `lzma` code into a generic compression function, so I added it, since it was quick and would allow use on non-binary files. I'm fine taking it out, if you'd prefer; it's a minor change.

Reply:
(In the commit before that, 8d554f9, I also replaced some wrong variable names, that were vestigial from before I'd wrapped the code into a function.)

Reply:
> would allow use on non-binary files

All files are binary in the sense that they're just a bunch of bytes. When you open a text file, you're opening a binary file that has an encoding which allows you to translate it to text.

If you run this function with `is_the_file_binary=True`, `shutil.copyfileobj` will throw an error:

```
TypeError: a bytes-like object is required, not 'str'
```

This is because `output_file_location` is always opened in binary mode with the current code. You could open it in text mode, and then Python would default to UTF-8 encoding. But this would potentially re-encode the text file.

So in short, when you use a utility to compress a file (like `gzip` or `xz`) it is always at the file (bytes) level and not at the text level.

So remove this and keep it simple!

Reply:
Ah, ok! Thank you for the explanation. Removed in b707e54!